### PR TITLE
[@types/lodash] Improve lodash

### PR DESCRIPTION
**Pull request by TonyPythoneer**
Please fill in this template.

- [ ] Prefer to make your PR against the `types-2.0` branch.
- [ ] Test the change in your own code.
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [ ] Run `npm run lint -- package-name` if a `tslint.json` is present.

If adding a new definition:
- [ ] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Run `tsc` without errors.
- [ ] Include the required [files](https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package) and header. Base these on the README, *not* on an existing project.

If changing an existing definition:
- [ ] Provide a URL to  documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.

### CODE REVIEW

**Comment by TonyPythoneer** [comment]
@andy-ms Would you take a look?
I follow your [advice](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/12528#issuecomment-261260404) on #12528. How do I improve this?

**Comment by andy-ms** [comment]
The module declarations should work if you remove the line `import * as _ from 'lodash';` and use the global `_` instead. See Microsoft/TypeScript#12348. You're also getting errors in the new tests because you're trying to assign to a const. New tests are probably not necessary since you've already tested these functions.

**Comment by TonyPythoneer** [comment]
@andy-ms Although this PR has passed Travis CI, the `lodash.xxx` is not  available as `lodash/xxx`.

Example:




I found `add1` is recognized as `string`, not `function`. Oppositely, `add2` haven't that issue, it even can show the hint normally. Please take a look. thx.

**Comment by andy-ms** [comment]
`export = typeof _.add` actually declares that you're exporting the result of the `typeof` operator (which returns a string).
Unfortunately you can't `export = _.add` because `_.add` is a method of `LodashStatic`, not a namespaced function.
So you'll have to do `const add: typeof _.add; export = add;`.

---

Another way would be to get rid of `LodashStatic` and directly use `export const` and `export function` inside of `declare module _`. Yhis could be left to another PR.

The line `declare var _: _.LoDashStatic;` would no longer be necessary.

The call signatures of `LodashStatic` would have to be declared *outside* like so:

```ts
declare function _(value: number): _.LoDashImplicitWrapper<number>;
declare function _(value: string): _.LoDashImplicitStringWrapper;
declare function _(value: boolean): _.LoDashImplicitWrapper<boolean>;
declare function _(value: Array<number>): _.LoDashImplicitNumberArrayWrapper;
declare function _<T>(value: Array<T>): _.LoDashImplicitArrayWrapper<T>;
declare function _<T extends {}>(value: T): _.LoDashImplicitObjectWrapper<T>;
declare function _(value: any): _.LoDashImplicitWrapper<any>;
```

Then inside `declare module _`, instead of:

```ts
interface LoDashStatic {
    add(
        augend: number,
        addend: number
    ): number;
}
```

You can do:

```ts
export function add(augend: number, addend: number): number;
```

Then `export = _.add` will work.

Also, `LodashStatic` is apparently used by `sequelize`, so you must replace `interface SequelizeLoDash extends _.LoDashStatic {` with `type SequelizeLoDash = typeof _ & {`.

**Comment by TonyPythoneer** [comment]
This latest commit is okay.

**Comment by TonyPythoneer** [comment]
@andy-ms Would you take a look? thx

File: lodash/modules.d.ts [hunk 1]
```diff
@@ -0,0 +1,1821 @@
...
+
+/**
+* uncoment it if definition exists
+*/
+/*
+declare module "lodash.pullAllWith" {
+   const pullAllWith: typeof _.pullAllWith;
+   export = pullAllWith;
+}
+*/
```
Comment (urielTwito) [inline lodash/modules.d.ts]: We're really missing the `pullAllWith` deceleration here on our project.
Why is this addition commented out?.. Any chance you can include it?..

**Comment by jasongrout** [comment]
This PR (now released as @types/lodash) breaks our repo (https://github.com/ipython/ipywidgets/tree/master/jupyterlab_widgets), which has lodash installed as a third-party dependency. We get errors like this when compiling now:

```
node_modules/@types/lodash/modules.d.ts(8,24): error TS2339: Property 'ary' does not exist on type 'UnderscoreStatic'.
node_modules/@types/lodash/modules.d.ts(20,29): error TS2339: Property 'assignIn' does not exist on type 'UnderscoreStatic'.
node_modules/@types/lodash/modules.d.ts(26,33): error TS2339: Property 'assignInWith' does not exist on type 'UnderscoreStatic'.
node_modules/@types/lodash/modules.d.ts(32,31): error TS2339: Property 'assignWith' does not exist on type 'UnderscoreStatic'.
node_modules/@types/lodash/modules.d.ts(38,23): error TS2339: Property 'at' does not exist on type 'UnderscoreStatic'.
node_modules/@types/lodash/modules.d.ts(62,28): error TS2339: Property 'bindKey' does not exist on type 'UnderscoreStatic'.
node_modules/@types/lodash/modules.d.ts(68,30): error TS2339: Property 'castArray' does not exist on type 'UnderscoreStatic'.
node_modules/@types/lodash/modules.d.ts(92,27): error TS2339: Property 'concat' does not exist on type 'UnderscoreStatic'.
node_modules/@types/lodash/modules.d.ts(135,26): error TS2339: Property 'curry' does not exist on type 'UnderscoreStatic'.
node_modules/@types/lodash/modules.d.ts(141,31): error TS2339: Property 'curryRight' does not exist on type 'UnderscoreStatic'.
node_modules/@types/lodash/modules.d.ts(159,33): error TS2339: Property 'defaultsDeep' does not exist on type 'UnderscoreStatic'.
node_modules/@types/lodash/modules.d.ts(183,33): error TS2339: Property 'differenceBy' does not exist on type 'UnderscoreStatic'.
node_modules/@types/lodash/modules.d.ts(189,35): error TS2339: Property 'differenceWith' does not exist on type 'UnderscoreStatic'.
```
and many more. Deleting the `/// <reference path="modules.d.ts" />` line in index.d.ts makes everything work again.

To reproduce:

```
git clone https://github.com/ipython/ipywidgets.git
cd ipywidgets/jupyterlab_widgets
npm install
```

`npm install` automatically runs the build which shows the errors. To run the build after installing, do `npm run build`.


I'll try to get a minimal example.

**Comment by jasongrout** [comment]
CC @SylvainCorlay ^

**Comment by andy-ms** [comment]
Hi @jasongrout, this should be fixed by #12995 which avoids using the global `_`.

**Comment by jasongrout** [comment]
Great, thanks @andy-ms!

**Pull request merged**